README.md
# LaunchKey SDK for Python - Web Example

  * [Overview](#overview)
  * [Pre-Requisites](#prerequisites)
  * [Installing](#installing)
  * [Usage](#usage)
  * [Support](#support)

<a name="overview"></a>
# Overview

LaunchKey is an identity and access management platform  This SDK enables developers to quickly integrate
the LaunchKey platform and Python based applications without the need to directly interact with the platform API.

Developer documentation for using the LaunchKey API is found [here](https://launchkey.com/docs/).

An overview of the LaunchKey platform can be found [here](https://launchkey.com/platform).

#  <a name="prerequisites"></a>Pre-Requisites

Utilization of the LaunchKey SDK requires the following items:

 * LaunchKey Account - The [LaunchKey Mobile App](https://launchkey.com/app) is required to set up a new account and
 access the LaunchKey Dashboard.

 * An application - A new application can be created in the [LaunchKey Dashboard](https://dashboard.launchkey.com/).
   From the application, you will need the following items found in the keys section of the application details:

    * The app key
    * The secret key
    * The private key

<a name="installing"></a>
# Installing

## Get the example code

```bash
git clone https://github.com/LaunchKey/launchkey-python-example-cli.git
```

or get a an archive from [the latest release](https://github.com/LaunchKey/launchkey-python-example-web/releases/latest).

## Install The Application

Once you have the code you need to install it:

```bash
$ python setup.py install
```

<a name="usage"></a>
# Usage

1. Get the app info necesary to start the web server

    * ```app_key``` - The app key value.  This value is automatically generate
    
    * ```secret_key``` - The secret key value.  Create a new one to copy if you have not done so already.
    
    * ```rsa_private_key_location``` - The location of the private key.  Download a new key from the dashboard if you have
    not done so already.  This must be an absolute path.

2. Start your web server

    ```bash
    $ launchkeywebdemo app_key secret_key rsa_private_key_location
      Starting server, use <Ctrl-C> to stop
    ```

3. Verify the server is running by accessing the URL of your web server: [http://127.0.0.1:8080](http://127.0.0.1:8080).

4. Start your reverse proxy.

    ```bash
    $ ngrok 8080
    ```

    Once started, you should see a a screen similar to:

    ```
    ngrok                                                                                          (Ctrl+C to quit)

    Tunnel Status                 online
    Version                       1.7/1.7
    Forwarding                    http://z4182320.ngrok.com -> 127.0.0.1:8080
    Forwarding                    https://z4182320.ngrok.com -> 127.0.0.1:8080
    Web Interface                 127.0.0.1:4040
    # Conn                        0
    Avg Conn Time                 0.00ms

    ```

4. Verify your reverse proxy by accessing the reverse proxy endpoint.  The endpoint will be the first part of one of the
Forwarding lines.  Based on the example above it would be ```https://z4182320.ngrok.com``` or
```http://z4182320.ngrok.com```.  Copy your value for the Forwarding endpoint into you browser to ensure it is
working correctly.  If working correctly, it will displaying the same web page you saw when verifying your web server
as well as show 200 OK responses in the HTTP Requests section of the ngrok screen like below:

    ```
    ngrok                                                                                          (Ctrl+C to quit)

    Tunnel Status                 online
    Version                       1.7/1.7
    Forwarding                    http://z4182320.ngrok.com -> 127.0.0.1:8080
    Forwarding                    https://z4182320.ngrok.com -> 127.0.0.1:8080
    Web Interface                 127.0.0.1:4040
    # Conn                        2
    Avg Conn Time                 3.07ms


    HTTP Requests
    -------------

    GET /favicon.ico              200 OK
    GET /                         200 OK

    ```

5. Now that your web server and reverse proxy are working, update your application with the callback URL.  This is done
by placing the URL you just verified from Ngrok into the callback field in the General section of your LaunchKey
application configuration.

5. __Winning!__ - You should be ready to try the demo and see how the Python SDK can be used to quickly and easily secure
your application with LaunchKey.

__NOTE:__ You will see numerous 401 log entries when logging in and a 403 or two when remote de-orbit occurs.  This is
expected as the javascript from the front end will poll for the status of the request.  The following are the expected
codes for the expected status:

* 200 - Authenticated
* 401 - Authenticating
* 403 - De-Orbited

<a name="support"></a>
# Support

## GitHub

Submit feature requests and bugs on [GitHub](https://github.com/LaunchKey/launchkey-python-example-cli/issues).

## Twitter

Submit a question to the Twitter Handle [@LaunchKeyHelp](https://twitter.com/LaunchKeyHelp).

## IRC

Engage the LaunchKey team in the `#launchkey` chat room on [freenode](https://freenode.net/).

## LaunchKey Help Desk

Browse FAQ's or submit a question to the LaunchKey support team for both
technical and non-technical issues. Visit the LaunchKey Help Desk [here](https://launchkey.desk.com/).


CHANGES.md
CHANGELOG for LaunchKey Python SDK Web Example
==============================================

1.1.1
-----

Bug fix - Add logout to server side deorbit handling

1.1.0
-----

Updates for cleaner code and distribution

* Use SQLite memory database to not have to write to the file system
* Move static files to their own directories at the root of the project
* Update the README to alert the user of why there are 401 and 403 responses in the logs
* Move the LaunchKey API instantiation to the main instead of the handler to reduce file access
* Override finish_request on the socket server to ignore broken pipe response errors when closing the connection
* Refactor handler do_POST do pull out duplicate code

1.0.0
-----

Initial release

webdemo/main.py
__copyright__ = 'Copyright 2015 LaunchKey, Inc.  See project license for usage.'
__author__ = 'Adam Englander (adam@launchkey.com)'

import sys, sqlite3, socket
from launchkey import API
from handlers import LaunchKeyHandler
from BaseHTTPServer import HTTPServer

class CleanServer(HTTPServer):
    def finish_request(self, request, client_address):
        '''
        Catch broken pipe errors for finish request.  They have no bearing here and simple upset people for no reason.
        :return:
        '''
        try:
            HTTPServer.finish_request(self, request, client_address)
        except socket.error, e:
            if e.errno == 32: # 32 is broken pipe
                pass
            else:
                raise e



def main():
    '''
    Main method that creates and runs app
    :param argv:
    :return:
    '''
    db_file = ':memory:'
    LaunchKeyHandler.sqlite = sqlite3.connect(db_file)
    LaunchKeyHandler.sqlite.execute('CREATE TABLE IF NOT EXISTS auth(request UNIQUE, status INT, timestamp INT, userhash)')

    args = sys.argv[1:]
    if len(args) != 3:
        print "\n\n"
        print "Usage:\n"
        print "    launchkeywebdemo app_key secret_key pk_location"
        print ""
        print "    app_key:     Your application's key"
        print "    secret_key:  Your application's secret key"
        print "    pk_location: The location of your application's RSA private key"
        print "\n\n"
        return

    app_key = args[0]
    secret_key = args[1]
    private_key_location = args[2]
    private_key = file(private_key_location).read()
    LaunchKeyHandler.launchkey = API(app_key, secret_key, private_key)

    server = CleanServer(('0.0.0.0', 8080), LaunchKeyHandler)
    print 'Starting server, use <Ctrl-C> to stop'
    return server.serve_forever()

if __name__ == '__main__':
    sys.exit(main())


webdemo/handlers.py
import socket

__author__ = 'adam'

from BaseHTTPServer import BaseHTTPRequestHandler


import os, urlparse, pystache, cgi, time
from pkg_resources import resource_filename

class LaunchKeyHandler(BaseHTTPRequestHandler, object):

    def do_GET(self):
        parsed_path = urlparse.urlparse(self.path)
        if parsed_path.path == '/favicon.ico':
            return self.__serve_favicon()

        auth_request = self.__get_auth_request_cookie()
        status = self.__get_status(auth_request)

        template = file(os.path.join(os.path.dirname(__file__), '..', 'templates', 'template.html')).read()

        context = {
            "unauthorized": True if status < 1 else None,
            "authorized": True if status == 2 else None,
            "authorizing": True if status == 1 else None,
        }

        page = pystache.render(template, context)
        self.send_response(200)
        self.send_header("Set-Cookie", "AuthRequest=%s" % (auth_request if auth_request is not None else ''))
        self.end_headers()
        self.wfile.write(page)
        self.close_connection = 1

    def do_POST(self):
        parsed_path = urlparse.urlparse(self.path)
        auth_request = self.__get_auth_request_cookie()
        form = cgi.FieldStorage(
            fp=self.rfile,
            headers=self.headers,
            environ={'REQUEST_METHOD':'POST', 'CONTENT_TYPE':self.headers.get('Content-Type'),}
        )

        query = urlparse.parse_qs(parsed_path.query)
        if form.has_key('username'):
            username = form.getvalue('username')
            auth_request = self.launchkey.authorize(username, True)
            status = 1
            self.sqlite.execute(
                'INSERT OR REPLACE INTO auth (request, status, timestamp) VALUES (?, ?, ?)',
                (auth_request, status, time.time())
            )
            self.sqlite.commit()
            self.send_response(302)
            self.send_header("Set-Cookie", "AuthRequest=%s" % (auth_request))
            self.send_header("location", "/")

        elif form.has_key('deorbit'):
            if auth_request:
                self.launchkey.logout(auth_request)
                self.sqlite.execute('DELETE FROM auth WHERE request = ?', (auth_request,))
                self.sqlite.commit()
            self.send_response(302)
            self.send_header("location", "/")

        elif query.has_key('auth') and query.has_key('auth_request') and query.has_key('user_hash'):
            auth_request = query.get('auth_request').pop()
            authorized = self.launchkey.is_authorized(auth_request, query.get('auth').pop())
            self.sqlite.execute(
                'UPDATE auth SET status = ?, userhash = ? where request = ?',
                (2 if authorized else 0, query.get('user_hash').pop(), auth_request)
            )
            self.sqlite.commit()
            self.send_response(200)

        elif query.has_key('deorbit') and query.has_key('signature'):
            user_hash = self.launchkey.deorbit(query.get('deorbit').pop(), query.get('signature').pop())
            if user_hash:
                code = 200
                c = self.sqlite.cursor()
                c.execute('SELECT request FROM auth WHERE userhash = ?', (user_hash,))
                for row in c.fetchall():
                    try:
                        request = row[0]
                        self.sqlite.execute('DELETE FROM auth WHERE request = ?', (request,))
                        self.sqlite.commit()
                        self.launchkey.logout(request)
                    except:
                        pass


            else:
                code = 400

            self.send_response(code)

        else:
            status = self.__get_status(auth_request)
            if status == 2:
                code = 200
            elif status == 1:
                code = 401
            else:
                code = 403

            self.send_response(code)

        self.end_headers()
        return

    def __serve_favicon(self):
        filename = resource_filename(__name__, '../static/favicon.ico')

        self.send_response(200)
        self.send_header('Content-Type', 'image/x-icon')
        self.send_header('Content-Length', os.stat(filename).st_size)
        self.end_headers()

        icofile = file(filename)
        while True:
            chunk = icofile.read(256)
            if chunk:
                self.wfile.write(chunk)
            else:
                icofile.close()
                return

    def __get_status(self, auth_request):
        status = 0
        if auth_request is not None and not auth_request == '':
            c = self.sqlite.cursor()
            c.execute('SELECT status FROM auth WHERE request = ? LIMIT 1', (auth_request,));
            row = c.fetchone()
            if row:
                (status,) = row
            c.close()
        return status

    def __get_auth_request_cookie(self):
        cookie_string = self.headers.getheader('Cookie')
        if cookie_string is not None:
            cookies = cookie_string.split(";")
            for cookie in cookies:
                (key, value) = cookie.strip().split("=")
                if key.strip() == 'AuthRequest':
                    return value

